src/Spanner/Session/CacheSessionPool.php

```diff
@@ -0,0 +1,548 @@
+            $data = (array) $item->get() ?: $this->initialize();
+
+            // If the queue has items in it, let's shift one off, however if the
+            // queue is empty and we have maxed out the number of sessions let's
+            // attempt to purge any orphaned items from the pool to make room
+            // for more.
+            if ($data['queue']) {
+                $session = $this->getSession($data);
+                $shouldSave = true;
+            } elseif ($this->config['maxSessions'] <= $this->getSessionCount($data)) {
+                $this->purgeOrphanedInUseSessions($data);
+                $this->purgeOrphanedToCreateItems($data);
+                $shouldSave = true;
+            }
+
+            $toCreate = $this->buildToCreateList($data, is_array($session));
+            $data['toCreate'] += $toCreate;
+
+            if ($shouldSave || $toCreate) {
+                $this->cacheItemPool->save($item->set($data));
+            }
+
+            return [$session, $toCreate];
+        });
+
+        // Create sessions if needed.
+        if ($toCreate) {
+            $createdSessions = [];
+            $exception = null;
+
+            try {
+                $createdSessions = $this->createSessions(count($toCreate));
+            } catch (\Exception $ex) {
+                $exception = $ex;
+            }
+
+            $session = $this->config['lock']->synchronize(function () use (
+                $session,
+                $toCreate,
+                $createdSessions,
+                $exception
+            ) {
+                $item = $this->cacheItemPool->getItem($this->cacheKey);
+                $data = $item->get();
+                $data['queue'] = array_merge($data['queue'], $createdSessions);
+
+                // Now that we've created the sessions, we can remove them from
+                // the list of intent.
+                foreach ($toCreate as $id => $time) {
+                    unset($data['toCreate'][$id]);
+                }
+
+                if (!$session && !$exception) {
+                    $session = array_shift($data['queue']);
+
+                    $data['inUse'][$session['name']] = $session + [
+                        'lastActive' => $this->time()
+                    ];
+                }
+
+                $this->cacheItemPool->save($item->set($data));
+
+                return $session;
+            });
+
+            if ($exception) {
+                throw $exception;
+            }
+        }
+
+        if ($session) {
+            $session = $this->handleSession($session);
+        }
+
+        // If we don't have a session, let's wait for one or throw an exception.
+        if (!$session) {
+            if (!$this->config['shouldWaitForSession']) {
+                throw new \RuntimeException('No sessions available.');
+            }
+
+            $session = $this->waitForNextAvailableSession();
+        }
+
+        return $this->database->session($session['name']);
```

Comment:
A single GAPIC client uses a single gRPC stub, which uses a single channel. So all requests on a GAPIC client will share that channel.